### ISSUE TITLE
Support fragment tracking for BPF DSR / SNAT with kube-proxy replacement

### ISSUE THREAD

**Issue by qmonnet** [cilium/cilium#11180]
Follow-up to #10264: Support fragment tracking with BPF-based DSR and SNAT.

Pretty sure `l4_modify_port()` and `l4_load_port()` in bpf/lib/l4.h, on SNAT path, should be changed to get fragment support. DSR probably needs other changes too.

The relevant metrics and tests need also be added.

**Comment by borkmann** [comment]
Also as discussed longer ago, this should have an agent flag on whether fragments should be handled from traffic coming into the node (bpf_host.o side). Should probably default to off and get documented in the kube-proxy guide (fragmentation handling for services).

**Comment by anfernee** [comment]
any updates @borkmann @qmonnet on this issue?

**Comment by qmonnet** [comment]
No, I don't believe anyone has had time to progress on this :/. Feel free to take it if you're interested.

**Comment by anfernee** [comment]
We would love to. It would be nice to have your suggestions. Will sync on slack on details.

**Comment by qmonnet** [comment]
Hi @anfernee, we never got back to this. Did you have a chance to make any progress? Or are you still planning to work on this? Please do feel free to reach on Slack if you want to discuss it.

**Comment by anfernee** [comment]
I think @liuyuan10 has spent time on this issue already. any updates?

**Comment by liuyuan10** [comment]
I didn't spend time on this issue. I was fixing the issue when kubeproxy is used.

**Comment by anfernee** [comment]
Got it. @qmonnet  What is the planned date for 1.10 release?

**Comment by brb** [comment]
@anfernee It's not clear yet, but my bet is Feb / beginning of March. Are you guys interested in working on this?

**Comment by anfernee** [comment]
Thanks @brb. I've synced with @qmonnet. We'll keep it in our plate, and we'll sync regularly on progress.

**Comment by anfernee** [comment]
Working on the DSR mode. I found when using dsr with `DSR_ENCAP_NONE`, the first packet is dropped. Tracing down to the following snippet:

```c
	if (ctx_adjust_room(ctx, 0x8, BPF_ADJ_ROOM_NET,
			    ctx_adjust_room_dsr_flags())) {
		return DROP_INVALID;
	}
```

This basically is calling into `skb_adjust_room`. I am wondering if this helper in kernel work with fragmented packets. Anyone knows?

**Comment by brb** [comment]
@anfernee Do you know the exact error code the skb_adjust_room returned?

**Comment by anfernee** [comment]
No. Good question! I will try it again tomorrow morning..

**Comment by pchaigno** [comment]
Could we be hitting [this check](https://elixir.bootlin.com/linux/v5.10.12/source/net/core/filter.c#L3637) in the kernel? If `skb->dev` is defined, then [we're bounded by](https://elixir.bootlin.com/linux/v5.10.12/source/net/core/filter.c#L3557) `skb->dev->mtu + skb->dev->hard_header_len` which is likely less than MTU + 8. I'm not sure when `skb-dev` is defined though (on TX? on RX once skb is assigned to dev?).

If that's the case, the error code should be `ENOTSUPP`.

**Comment by anfernee** [comment]
@pchaigno you are right. it is `-524`
```
#define ENOTSUPP	524	/* Operation is not supported */ 
```

**Comment by borkmann** [comment]
> Could we be hitting [this check](https://elixir.bootlin.com/linux/v5.10.12/source/net/core/filter.c#L3637) in the kernel? If `skb->dev` is defined, then [we're bounded by](https://elixir.bootlin.com/linux/v5.10.12/source/net/core/filter.c#L3557) `skb->dev->mtu + skb->dev->hard_header_len` which is likely less than MTU + 8. I'm not sure when `skb-dev` is defined though (on TX? on RX once skb is assigned to dev?).

In both ingress + egress hooks it must be set.

**Comment by borkmann** [comment]
> I found when using dsr with DSR_ENCAP_NONE, the first packet is dropped.
Is this due to TFO?

**Comment by brb** [comment]
> Is this due to TFO?

Or UDP?

**Comment by anfernee** [comment]
> > Is this due to TFO?
> 
> Or UDP?

TCP Fast Open? No, I was using UDP. I was running a UDP echo service in the cluster, and sent a 2k file (bigger than MTU) to that service. What I found was interesting: the first packet was dropped, but the second packet was delivered.

**Comment by pchaigno** [comment]
> I was running a UDP echo service in the cluster, and sent a 2k file (bigger than MTU) to that service. What I found was interesting: the first packet was dropped, but the second packet was delivered.

First packet would be MTU sized, second packet would be smaller, right? So the `ENOTSUPP` error is expected. Not sure how we can work around that though :-/

**Comment by brb** [comment]
@anfernee Can you control the client-side MTU?

**Comment by anfernee** [comment]
> @anfernee Can you control the client-side MTU?

Yes, I can configure it on the network interface from client-side.

**Comment by brb** [comment]
> Yes, I can configure it on the network interface from client-side.

Lowering the client MTU by 8 bytes for the DSR mode with the IPv4 option should make it work.

**Comment by borkmann** [comment]
> > @anfernee Can you control the client-side MTU?
> 
> Yes, I can configure it on the network interface from client-side.

Also I'm getting a PR out next days that would cause an ICMP reply with frag needed returning `MTU - encap_ohead`. Maybe that would also address this case for clients as a fall-back..

**Comment by rogerBridge** [comment]
Is there a way to change server-side settings(like MTU) to solve this question?

**Comment by janz0390** [comment]
@julianwiedmann is your fix in #26550 supposed to fully address this issue too?

**Comment by julianwiedmann** [comment]
> @julianwiedmann is your fix in #26550 supposed to fully address this issue too?

Not that I'm aware. I would expect that more work is needed. Starting with tests :).
